### PR TITLE
fix: use idScheme in event dataElement checks DHIS2-12282

### PR DESCRIPTION
**Pull request by teleivo**
when a user uses a dataElementIdScheme other than UID the NTI validation falsely said
* the payload is missing a mandatory dataElement (mandatory to the program stage)
* the payload contains a dataElement that is not part of the events programStage

This was due to the NTI validation not taking the idScheme into account in the EventDataValuesValidationHook

After the above fixes the ProgramRuleService failed to find the DataElement because it was using the UID instead of the user specified CODE idScheme.

Now the EventTrackerConverterService map EventDataValue.dataElement to UID. It does so by getting the dataElement from the preheat.

At this point EventDataValuesValidationHook already checked whether the
dataElement specified by the user exists by looking in the preheat. If
not found it would add error E1304 'DataElement `{0}` is not a valid
data element' and fail validation. Looking the dataElement up in the
preheat already respects the idScheme.

Lookup the dataElement in the preheat using the user specified identifier.
Use the UID when mapping to an EventDataValue. This means any code using
the EventDataValue.identifier from then on can rely on it being a UID.
The NTI request idScheme is not available in the program rule engine so
any existing code already assumed it was a UID.

* fixes a bug in EventTrackerConverterService that fails to merge new and existing dataValues if the user uses an idScheme other than UID.
* EventDataValuesValidation and EventTrackerConverterServiceTest hook test now uses mockitos strict mode. I removed unnecessary mockito stubbing so tests remain clean in the future. Extracted helpers to make differences between tests clear.

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/converter/EventTrackerConverterServiceTest.java [hunk 1]
```diff
@@ -27,16 +27,19 @@
  */
 package org.hisp.dhis.tracker.converter;
 
+import static org.hisp.dhis.utils.Assertions.assertContainsOnly;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.mockito.Mockito.when;
+import static org.wildfly.common.Assert.assertTrue;
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/converter/EventTrackerConverterServiceTest.java:34]: i will switch that one to a junit assertion.

File: dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/converter/EventTrackerConverterServiceTest.java [hunk 2]
```diff
@@ -170,6 +170,129 @@ void testToProgramStageInstance()
         } );
     }
 
+    @Test
+    void fromForRuleEngineGivenNewEvent()
```
Comment (zubaira) [inline dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/converter/EventTrackerConverterServiceTest.java:174]: fromFor ?

**Review by zubaira — APPROVED** [review]
just a small comment, otherwise LGTM.

File: dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/converter/EventTrackerConverterServiceTest.java (hunk 2 above)
Comment (teleivo) [reply dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/converter/EventTrackerConverterServiceTest.java:174]: `fromForRuleEngine` is the name of the method I am testing 😁
defined in RuleEngineConverterService

**Review by enricocolasante — APPROVED** [review]

**Pull request merged**